File: runtime/CSharp/Antlr4.Runtime/Atn/OrderedATNConfigSet.cs
// Copyright (c) Terence Parr, Sam Harwell. All Rights Reserved.
// Licensed under the BSD License. See LICENSE.txt in the project root for license information.

using Antlr4.Runtime.Sharpen;

namespace Antlr4.Runtime.Atn
{
    /// <author>Sam Harwell</author>
    public class OrderedATNConfigSet : ATNConfigSet
    {
        public OrderedATNConfigSet()
        {
        }

        public OrderedATNConfigSet(ATNConfigSet set, bool @readonly)
            : base(set, @readonly)
        {
        }

        public override ATNConfigSet Clone(bool @readonly)
        {
            Antlr4.Runtime.Atn.OrderedATNConfigSet copy = new Antlr4.Runtime.Atn.OrderedATNConfigSet(this, @readonly);
            if (!@readonly && this.IsReadOnly)
            {
                copy.AddAll(this);
            }
            return copy;
        }

        protected internal override long GetKey(ATNConfig e)
        {
            return e.GetHashCode();
        }

        protected internal override bool CanMerge(ATNConfig left, long leftKey, ATNConfig right)
        {
            return left.Equals(right);
        }
    }
}

File: runtime/Java/src/org/antlr/v4/runtime/atn/OrderedATNConfigSet.java
/*
 * Copyright (c) 2012 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD-3-Clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

package org.antlr.v4.runtime.atn;

/**
 *
 * @author Sam Harwell
 */
public class OrderedATNConfigSet extends ATNConfigSet {

	public OrderedATNConfigSet() {
	}

	public OrderedATNConfigSet(ATNConfigSet set, boolean readonly) {
		super(set, readonly);
	}

	@Override
	public ATNConfigSet clone(boolean readonly) {
		OrderedATNConfigSet copy = new OrderedATNConfigSet(this, readonly);
		if (!readonly && this.isReadOnly()) {
			copy.addAll(this);
		}

		return copy;
	}

	@Override
	protected long getKey(ATNConfig e) {
		return e.hashCode();
	}

	@Override
	protected boolean canMerge(ATNConfig left, long leftKey, ATNConfig right) {
		return left.equals(right);
	}

}
